### DISCUSSION TITLE
[Help]: OC200 Controller Migrate to New Docker Controller Error

### DISCUSSION THREAD

**Discussion by BillyBobMills** [mbentley/docker-omada-controller#376] [Help] [answered]
### Controller Version

v5.12.7

### Describe Your Issue or Question

Putting this in its own discussion to separate this issue. 

When trying to restore a controller backup or using the controller migration steps from an OC200 running v5.12.9 into a new Proxmox/Ubuntu 22.04 VM/docker software controller v5.12.7 instance I get a UI error. "Restore Failed, Unknown error" 

A site only backup from the OC200 will restore into the docker software controller successfully.

A log error is below. Is this a permission problem?

I am running the container as a user with id 1000 group id 1000 who is in the docker group.
user2@ubuntutest2:~$ id
uid=1000(user2) gid=1000(user2) groups=1000(user2),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),117(lxd),998(docker)

Am I missing the point with the 508 uid and guid from the docs?

I can successfully restore the same controller backup to a fresh Ubuntu 20.04 VM which I have installed the omada deb file using the below commands. So the backup file is not the issue.
sudo apt-get install mongodb jsvc openjdk-8-jdk gdebi-core -y
sudo update-alternatives --config java > option 2
wget https://static.tp-link.com/upload/software/2023/202309/20230920/Omada_SDN_Controller_v5.12.7_Linux_x64.deb
sudo gdebi Omada_SDN_Controller_v5.12.7_Linux_x64.deb

Additionally to rule out Ubuntu 22.04 vs 20.04 I have tried the docker container on another fresh 20.04 ubuntu / docker install.

I have tried the docker image using a stack in portainer and via a docker command to rule that out. 

I can only imagine I am missing something with the Docker Container process?

### Expected Behavior

Migration / Restore is successful

### Steps to Reproduce

Backup OC200 controller
Restore Controller in new Docker Software Controller
Error on restore

### How You're Launching the Container

```plain
docker run -d \
  --name omada-controller \
  --restart unless-stopped \
  --ulimit nofile=4096:8192 \
  -p 8088:8088 \
  -p 8043:8043 \
  -p 8843:8843 \
  -p 27001:27001/udp \
  -p 29810:29810/udp \
  -p 29811-29816:29811-29816 \
  -e PGID="1000" \
  -e PUID="1000" \
  -e MANAGE_HTTP_PORT=8088 \
  -e MANAGE_HTTPS_PORT=8043 \
  -e PORTAL_HTTP_PORT=8088 \
  -e PORTAL_HTTPS_PORT=8843 \
  -e PORT_ADOPT_V1=29812 \
  -e PORT_APP_DISCOVERY=27001 \
  -e PORT_DISCOVERY=29810 \
  -e PORT_MANAGER_V1=29811 \
  -e PORT_MANAGER_V2=29814 \
  -e PORT_TRANSFER_V2=29815 \
  -e PORT_RTTY=29816 \
  -e PORT_UPGRADE_V1=29813 \
  -e SHOW_SERVER_LOGS=true \
  -e SHOW_MONGODB_LOGS=false \
  -e SSL_CERT_NAME="tls.crt" \
  -e SSL_KEY_NAME="tls.key" \
  -e TZ=Etc/UTC \
  -v /home/user2/config/omada/data:/opt/tplink/EAPController/data \
  -v /home/user2/config/omada/logs:/opt/tplink/EAPController/logs \
  mbentley/omada-controller:latest
```

### Container Logs

```plain
11-28-2023 13:57:55.992 ERROR [restore-work-group-0] [] c.t.s.o.b.c.r.c(): omadacId=OmadacId(XXX) restore failed with Exception
java.lang.reflect.InaccessibleObjectException: Unable to make field private static final long java.util.LinkedHashMap.serialVersionUID accessible: module java.base does not "opens java.util" to unnamed module @5c130274
at java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354) ~[?:?]
at java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297) ~[?:?]
at java.lang.reflect.Field.checkCanSetAccessible(Field.java:178) ~[?:?]
at java.lang.reflect.Field.setAccessible(Field.java:172) ~[?:?]
at com.tplink.smb.omada.backup.api.f.c.a(SourceFile:27) ~[backup-api-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.api.f.b.a(SourceFile:54) ~[backup-api-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.core.unit.a.a.a(SourceFile:31) ~[backup-core-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.core.restore.c.a(SourceFile:606) ~[backup-core-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.core.restore.c.a(SourceFile:538) ~[backup-core-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.core.restore.c.a(SourceFile:181) ~[backup-core-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.core.restore.j.a(SourceFile:124) ~[backup-core-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.core.restore.j.a(SourceFile:65) ~[backup-core-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.core.a.a.a(SourceFile:100) ~[backup-core-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.a.b.a.a.f.b(SourceFile:230) ~[maintenance-port-web-5.12.7.jar:5.12.7]
at com.tplink.smb.component.lock.api.AbstractReadWriteLockService.doWithTryLockWrite(AbstractReadWriteLockService.java:107) ~[solution-components-lock-api-1.1.4.jar:1.1.4]
at com.tplink.smb.omada.lock.a.a.b(SourceFile:77) ~[omada-lock-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.lock.a.a.b(SourceFile:92) ~[omada-lock-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.a.b.a.a.f.a(SourceFile:227) ~[maintenance-port-web-5.12.7.jar:5.12.7]
at com.tplink.smb.omada.backup.a.b.a.a.f.a(SourceFile:161) ~[maintenance-port-web-5.12.7.jar:5.12.7]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
at java.lang.Thread.run(Thread.java:833) [?:?]
11-28-2023 13:57:55.995 INFO [restore-work-group-0] [] c.t.s.o.b.c.r.b.a(): omadacId=XXX restore failed
```

**Comment by mbentley** [comment]
So I'm sorry but it is not clear which all of the scenarios you described fail and which do succeed as I was wanting to understand if there is potentially some issue with some software package missing in the container or something else.  Could you list what were the results of each of the scenarios you described on a pass/fail?

**Comment by BillyBobMills** [comment]
Yes sorry it was a bit of a ramble

Pass:
OC200 Site backup > Ubuntu 22.04 VM with mbentley/omada-controller:latest (via Portainer stack)
OC200 Controller backup > Ubuntu 20.04 VM with Omada_SDN_Controller_v5.12.7_Linux_x64.deb installed.

Fail:
OC200 Controller backup > Ubuntu 22.04 VM with mbentley/omada-controller:latest (via Portainer stack and docker run)
OC200 Controller backup > Ubuntu 20.04 VM with mbentley/omada-controller:latest (via Portainer stack and docker run)

**Reply by mbentley** [reply]
OK, cool, so it seems that anything related to a full _controller_ backup and restore is failing.  I will see if I can replicate this just with a controller in a container backup and restore to a new one.

**Reply by mbentley** [reply]
OK, so I took a backup and was able to directly restored that configuration to a new container.  Are your failures when you're doing a _migration_ or is it when you're taking a file based backup and doing a restore?  I have not tried any migrations (yet).

**Comment by BillyBobMills** [comment, marked as answer]
Hi
Thanks for looking into this.
I had tried both a migration and a backup/restore. (and tried settings only, with and without user data, and various lengths of data backed up).
The migration does what appears to be just a backup which you then restore onto the new controller.
A site migration seems to be the only way to "backup" a "site" as there is no separate way to do that. 

I have since got this working on the new controller by restoring the OC200 backup to the ubuntu 20.04 .deb install, taking a backup from there and then restoring that into the container instance. All seems to be working ok. 


Pass:
OC200 Controller backup > Ubuntu 20.04 VM with Omada_SDN_Controller_v5.12.7_Linux_x64.deb installed > Ubuntu 22.04 VM with mbentley/omada-controller:latest (via Portainer stack with ports defined)

The OC200 backup I used was: 365 day, retain user info, to local file. I did the same backup on the ...Linux_x64.deb install which was restored onto the container instance.

I did not migrate the devices to the ...Linux_x64.deb install. I used it only to perform a restore and then create a new backup file. 

Once the backup was restored onto the new controller all devices showed as "Disconnected". After turning off the OC200 the devices on the new controller began to adopt, provision and then connect.

The known clients are maintained at the controller not site level and the length of data retained in the backup has an impact on this list (clients last seen x days ago). 

Working Portainer Stack / Docker Compose:

version: "3.1"

services:
  omada-controller:
    container_name: omada-controller
    image: mbentley/omada-controller:latest
    restart: unless-stopped
    ulimits:
      nofile:
        soft: 4096
        hard: 8192
    stop_grace_period: 60s
    ports:
      - 8088:8088
      - 8043:8043
      - 8843:8843
      - 27001:27001/udp
      - 29810:29810/udp
      - 29811-29816:29811-29816
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Etc/UTC
    volumes:
      - /home/user1/config/omada/data:/opt/tplink/EAPController/data
      - /home/user1/config/omada/logs:/opt/tplink/EAPController/logs

Thanks again for assisting with my issue and for creating and maintaining the docker image.
